clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java

```diff
@@ -346,4 +347,85 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
      * @return                The AlterConfigsResult
      */
     public abstract AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs, AlterConfigsOptions options);
+
+    /**
+     * Change the log directory for the specified replicas. This API is currently only useful if it is used
+     * before the replica has been created on the broker. It will support moving replicas that are already created after
+     * KIP-113 is fully implemented.
+     *
+     * Updates are not transactional so they may succeed for some resources while fail for others. The log directory for
+     * a particular replica is updated atomically.
+     *
+     * This operation is supported by brokers with version 0.11.1.0 or higher.
+     *
+     * @param replicaAssignment  The replicas with their log directory absolute path
+     * @return                   The AlterReplicaDirResult
+     */
+    public AlterReplicaDirResult alterReplicaDir(Map<TopicPartitionReplica, String> replicaAssignment) {
```

Comment:
Not sure if we want to have this interface added while we have not supported the intra broker replica move yet. Also, we probably don't want to expose the implementation detail to the users about how the replicas are moved . So it would be better to not ask users to manually specify the replica dir before they do the partition reassignment.

So the end state of AdminClient after KIP-179 and KIP-113 would be having two methods for partition movement:
1. A method for partition reassignment in general, include both inter and intra broker reassignment. (e.g. `alterTopics()`)
2. A method to only move replicas within a broker. (e.g. `alterReplicaDir`)

BTW, I still feel that having (1) is sufficient. It would be useful to explore possibility of only having (1) so users don't need to deal with two different interfaces for replica movement.

With that, regarding this patch, I was thinking doing the following:
1. Let `ReassignPartitionCommand` take new input format which includes the log dir.
2. Do a sanity check in the `ReassignParititionCommand` to ensure no intra broker replica movement is specified. Otherwise throw `UnsupportedOperationException`.
3. In `ReassignPartitionCommand`, it should just send `ChangeReplicaDirRequest` to the related brokers before the partition movement.
4. Do not add `alterReplicaDir()` to the AdminClient until the broker supports that.

Would the above way be sufficient for what we want to achieve for this patch?

Reply:
Sorry for being late on this. Why is this `alterReplicaDir` instead of `alterReplicaDirs`? This is a batch api like every other API so it should indicate that via the name, right?

Reply:
@ijuma I think one reason to use alterReplicaDir is that it presents a map from replica -> dir. It is probably a bit different from other names such as describeConfigs, which represents a collection of configs.

Do you like me to submit a patch to change it to a different name? @junrao @becketqin What do you think about this name?

Reply:
Thanks for the quick reply. Right, it's a map from replica to dir instead of a single replica to dir. The name sounds like the latter to me. Before doing a PR, let's see if we get consensus amongst yourself and the reviewers.